C++: Add an experimental query for surprising lifetimes from range-based for loops

MathiasVP opened the pull request:
This PR adds an experimental query to detect issues such as:

```cpp
std::vector<std::vector<int>> CreateCollection();

void use(...);

void test() {
  // BUG: CreateCollection() is destroyed before CreateCollection()[0].begin() is called
  for (auto x : CreateCollection()[0]) {
    use(x);
  }
}
```

I think the precision of this query is actually perfectly suited for Code Scanning (🤞), but since the query depends on ongoing work on both the extractor side and the [analysis side](https://github.com/github/codeql/pull/15858) it'll start its lifetime (pun very much intended!) in experimental.

All the `NOT DETECTED` tests are actually detected when running with the yet-to-be-released version of the extractor and with https://github.com/github/codeql/pull/15858 merged.

I don't think it makes much sense to run this on DCA yet as that DCA run will mostly test https://github.com/github/codeql/pull/15858 (and the internal extractor changes) I guess. I'll try doing it anyway just for fun, though!

geoffw0 commented on cpp/ql/test/experimental/query-tests/Security/CWE/CWE-416/test.cpp:
```diff
@@ -0,0 +1,719 @@
+	template<class T> constexpr T&& forward(remove_reference_t<T>&& t) noexcept;
+}
+
+// --- iterator ---
```
Is there anything stopping us moving some of these headers into common files (shared across all tests that need them)?  It's something that's come up before, but the particularly large number of `std` headers in this test is making me think this improvement might be due.

MathiasVP replied:
We should totally work towards that, yes. I'm not quite sure how to do this since each of our tests live in separate directories with no common root so, as far as I know, we can't just `#include "some-directory.h"`

geoffw0 replied:
Once upon a time we could include files outside the test directory, then we couldn't, at this point I genuinely don't know.

geoffw0 replied:
I've written this up as an issue, since it was bothering me.

MathiasVP replied:
Thanks!

geoffw0 commented on cpp/ql/test/experimental/query-tests/Security/CWE/CWE-416/test.cpp:
```diff
@@ -0,0 +1,719 @@
+
+  for (auto x : returnRef()) {} // GOOD
+  for (auto x : returnRef()[0]) {} // GOOD
+  for (auto x : returnRef().at(0)) {} // GOOD
```
Inconsistent indentation here (not that it really matters in a test).

MathiasVP replied:
Fixed it in 668239f355ea11d21e05fbe272675baca35fb547 anyway

geoffw0 commented on cpp/ql/test/experimental/query-tests/Security/CWE/CWE-416/test.cpp:
```diff
@@ -0,0 +1,719 @@
+
+void test() {
+  for (auto x : returnValue()) {} // GOOD
+  for (auto x : returnValue()[0]) {} // BAD [NOT DETECTED] (see *)
```
I notice the `BAD` cases don't contain a use of `x` inside the loop, like in the example code. 
 I guess the very act of iterating through the (released temporary) object causes undefined behaviour.

MathiasVP replied:
Yeah, I didn't bother complicating the query by actually looking for a use of the variable. If you're using range-based for loop the iterator will already have been dereferenced (see the Explanation section [here](https://en.cppreference.com/w/cpp/language/range-for) for how it's desugared).

geoffw0 commented on cpp/ql/test/experimental/query-tests/Security/CWE/CWE-416/test.cpp:
```diff
@@ -0,0 +1,719 @@
+  for (auto x : returnRef()[0]) {} // GOOD
+  for (auto x : returnRef().at(0)) {} // GOOD
+
+	for(auto it = returnValue().begin(); it != returnValue().end(); ++it) {} // BAD
```
This is also bad because the `begin()` and `end()` iterators aren't from the same object, right?

MathiasVP replied:
Indeed

geoffw0 commented on cpp/ql/test/experimental/query-tests/Security/CWE/CWE-416/test.cpp:
```diff
@@ -0,0 +1,719 @@
+
+	for (auto x : return_self_by_ref(returnValue())) {} // BAD [NOT DETECTED] (see *)
+
+	for (auto x : return_self_by_value(returnValue())) {} // GOOD
```
How confident are you that each of these test cases are annotated correctly?  Would it be worth one of us trying to prove actual failures for each case?

MathiasVP replied:
I tested them all on https://ideone.com/ before pushing these annotations 😄

MathiasVP replied:
And the annotations also matches my understanding of the semantics of temporary objects, at least. So that's also a bonus!

geoffw0 replied:
Awesome, I appreciate the thoroughness!

geoffw0 commented on cpp/ql/src/experimental/Security/CWE/CWE-416/IteratorToExpiredContainer.ql:
```diff
@@ -0,0 +1,89 @@
+/**
+ * @name Iterator to expired container
+ * @description Using an iterator owned by a container whose lifetimes has expired may lead to unexpected behavior.
```
```suggestion
 * @description Using an iterator owned by a container whose lifetime has expired may lead to unexpected behavior.
```

geoffw0 commented on cpp/ql/src/experimental/Security/CWE/CWE-416/IteratorToExpiredContainer.ql:
```diff
@@ -0,0 +1,89 @@
+}
+
+/**
+ * Flow from any destroyed object to the qualifier of a `begin` call
```
```suggestion
 * Flow from any destroyed object to the qualifier of a `begin` or `end` call
```
Alternatively we could limit the sink to a "begin" function.

geoffw0 commented on cpp/ql/src/experimental/Security/CWE/CWE-416/IteratorToExpiredContainer.ql:
```diff
@@ -0,0 +1,89 @@
+  predicate isSource(DataFlow::Node source) { source = getADestroyedNode() }
+
+  predicate isSink(DataFlow::Node sink) { isSinkImpl(sink, _) }
+}
```
What stops use getting a result when the sequence of events is:

1. create temporary object
2. iterate through temporary object
3. delete temporary object

MathiasVP replied:
How do you suppose we iterate through the temporary object before it's destroyed? 🤔

MathiasVP replied:
Oh, wait. Maybe I can think of something ❗

MathiasVP replied:
Yep, I managed to come up with something in https://github.com/github/codeql/pull/15939/commits/e373341f62a133b45a1f86cdc8f730d81d4c468f (see line 747 in that commit). I've fixed the FP in b944f3b411f31c704db467696ac8c494bf2ab714.

geoffw0 replied:
Changes LGTM. :+1:

geoffw0 reviewed (commented):
This looks like a great query, and the `.qhelp` is a good explanation of the problem (with references for more detail).  :+1:

Some questions below.

We also need to do a DCA run, docs review, and a closer examination of real world results ... but for each of these, I don't really mind whether we do it now or when we're ready to promote the query out of experimental.  It would certainly help to have the extractor change merged.

MathiasVP commented:
> We also need to do a DCA run, docs review, and a closer examination of real world results ... but for each of these, I don't really mind whether we do it now or when we're ready to promote the query out of experimental. It would certainly help to have the extractor change merged.

Partial answer: See what I wrote above:
> I don't think it makes much sense to run this on DCA yet as that DCA run will mostly test https://github.com/github/codeql/pull/15858 (and the internal extractor changes) I guess. I'll try doing it anyway just for fun, though!

subatoi reviewed (dismissed):
👍 from Docs

MathiasVP commented:
I believe I've addressed all your comments now @geoffw0. In terms of
> We also need to do a DCA run, docs review, and a closer examination of real world results ... but for each of these, I don't really mind whether we do it now or when we're ready to promote the query out of experimental. It would certainly help to have the extractor change merged.

I believe it's best merge this query, and then we can do proper MRVA result investigations and DCA tests once we move the query out of experimental (since, at that point, we'll have both the QL and extractor changes merged, and possibly even MRVA data depending on how lucky we are with the timing of a release 🤞)

geoffw0 reviewed (approved):
and :+1: from me.
